### PR TITLE
feat: Simulated KV cache QARL

### PR DESCRIPTION
**Pull request by mxinO**
# What does this PR do ?

Add simulated KV cache quantization for QARL

# Issues
* This enables us to experiment different kv cache quant configs with modelopt. e.g. nvfp4, skip first/last N tokens etc.
* A real quant of NVFP4 kv cache and maybe FP8 should be after this PR.
* Note that nemo-RL already has a FP8 kvcache real-quant support with per-step calibration, but not using the modelopt path. I will check if we can unify to one instead of splitting things.

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* depends on https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4591
* **Note this updated the MBridge commit to include the MBridge PR.** Not verified whether it breaks other dependencies.

# Reward Curves

Result-bearing Llama-3.1-8B QA-GRPO runs are shown below. All arms use the same rollout and training configuration; only simulated K/V quantization changes. The two optimizer learning rates are plotted separately. Faint lines are raw training reward and solid lines are an EWMA with span 20.

KV-cache QARL training reward

KV-cache QARL validation accuracy

The 3e-7 comparison includes BF16, calibrated FP8, constant-amax FP8, and calibrated NVFP4. The 1e-6 comparison includes BF16, calibrated FP8, and calibrated NVFP4. These are single-run curves, and arms stopped at different steps.

## Rollout Diagnostics

The same audited runs are compared using response length, rollout/policy mismatch KL, and policy entropy.

KV-cache QARL rollout diagnostics

## Conclusion

In the matched 3e-7 experiment, constant-amax FP8 performed better than calibrated FP8. Constant amax uses amax=448, so the E4M3 quantization scale is amax / 448 = 1. Its final-20 mean training reward was 0.5139 versus 0.4983 for calibrated FP8, and its final observed validation accuracy was 0.5547 versus 0.5352. This is a single-run result; constant-amax FP8 was not included in the 1e-6 comparison.

### CODE REVIEW

**Comment by realAsma** [comment]
LGTM overall,
@kinjalpatel27  Could you also take a look at this PR?

File: docs/assets/qarl_kv_cache_validation_accuracy.png
Comment (sharonyu-115) [inline docs/assets/qarl_kv_cache_validation_accuracy.png:1]: Hi @mxinO, The two KV-cache curve images added in this PR are not referenced by this guide or anywhere else in the tree so they will not appear in the generated docs. Could you either embed both curves in the doc with explanatory text, or remove to have them in the PR only?

Comment (sharonyu-115) [reply docs/assets/qarl_kv_cache_validation_accuracy.png:1]: Besides, I wonder if you can also share the other key curves including response length, mismatch kl, entropy, etc.?

File: docs/guides/quantization-aware-rl.md [hunk 1]
```diff
@@ -30,6 +30,38 @@ The following workflow + quantization recipe combinations have been validated en
...
+- `kv_cache_nvfp4.yaml`: calibrated NVFP4 K/V fake quantization with dynamic
+  per-block scales and a calibrated global amax.
+
+These recipes intentionally disable weight, input, Q, and P quantizers so a
+test can isolate simulated K/V behavior. They can also serve as the K/V portion
+of a complete user-authored recipe. Constant-amax K/V recipes need no amax
+transfer, but policy and rollout must still use the same recipe.
+
+This path does not enable vLLM's native FP8 KV-cache storage. Set neither
+`policy.generation.real_quant` nor a native vLLM KV-cache dtype for these
```
Comment (sharonyu-115) [inline docs/guides/quantization-aware-rl.md:69]: It would be good to add guard in the code for the constraints mentioned in the doc here: rollout and training have to use the same quant config and no real_quant & vLLM KV-cache dtype.

File: tests/unit/models/generation/test_vllm_quant_backend.py [hunk 2]
```diff
@@ -109,18 +119,14 @@ def cluster():
...
-def test_vllm_quant_refit(cluster, async_engine):
-    """Integration test: quantized Megatron -> vLLM refit should transfer pre-folded weights.
-
-    Uses is_eval=True so vLLM loads real HF weights at init, allowing us
-    to verify that pre-folded weights (with FP8 quantization applied on
-    the Megatron side) differ from the original HF weights by a small
-    quantization error.
-    """
+@pytest.mark.parametrize("recipe", ["kv_cache_fp8.yaml", "kv_cache_nvfp4.yaml"])
+def test_vllm_quant_refit(cluster, recipe, monkeypatch):
```
Comment (sharonyu-115) [inline tests/unit/models/generation/test_vllm_quant_backend.py]: It seems like this test is now repurposed to verify the new kv amax fidelity. The original weights refit test is replaced. would it be better to have a new `test_vllm_kv_quant_refit` instead of replacing the existing test?

Comment (mxinO) [reply tests/unit/models/generation/test_vllm_quant_backend.py]: Restored the original `test_vllm_quant_refit` coverage, including both sync and async engines and the pre-folded weight assertions. Added a separate parameterized `test_vllm_kv_quant_refit` for the FP8 and NVFP4 K/V recipes, which compares calibrated K/V amax state after Megatron-to-vLLM refit.

File: docs/assets/qarl_kv_cache_validation_accuracy.png
Comment (mxinO) [reply docs/assets/qarl_kv_cache_validation_accuracy.png:1]: Added a rollout-diagnostics figure to the PR description using the same audited runs. It reports response length (`train/mean_gen_tokens_per_sample`), rollout/policy mismatch KL (`train/gen_kl_error`), and policy entropy (`train/approx_entropy`). The figure is hosted externally so experiment images are not included in the repository diff.

File: docs/guides/quantization-aware-rl.md (hunk 1 above)
Comment (mxinO) [reply docs/guides/quantization-aware-rl.md:69]: You are right. Policy and rollout should use the same quantization recipe so calibrated amax state is transferred consistently and both paths use matched quantization. The supported exception is rollout-only FP8 K/V fake quantization with `use_constant_amax: true`; vLLM computes that constant amax locally, so it does not require policy-side K/V quantization or amax transfer.

Comment (sharonyu-115) [reply docs/guides/quantization-aware-rl.md:69]: Hi @mxinO, I see the doc has been updated to allow enabling K/V
quantization on only one side. I have questions on how this will work for the following cases: 
1. Calibrated amax +  K/V quantization on the rollout side only: If I understand correctly, vLLM initializes amax to -1 expecting refit will overwrite the value https://github.com/NVIDIA-NeMo/RL/blob/30366c4883e0287176c460f06956cde9c0a23830/nemo_rl/modelopt/models/generation/vllm_quant_patch.py#L233-L238. If training side does not have K/V quantization enabled, what will happen? Will this silently fallback to static amax or issue an error?
2. Static amax + K/V quantization on the rollout side only: this seems to be a working path. 
3. K/V quantization on the training side only: in this case amax will be pass to vLLM during refit without a destination on vLLM side. What will be the behavior under this case?

File: docs/assets/qarl_kv_cache_validation_accuracy.png
Comment (sharonyu-115) [reply docs/assets/qarl_kv_cache_validation_accuracy.png:1]: Hi @mxinO, Thanks! Looking at the curves, it seems like the NVFP4 KV quantization run has some numerics and convergence issue - very high gen_kl_error and gap on acc with other runs. I wonder if this is a valid recipe. What is your opinion on this?

File: examples/modelopt/quant_configs/kv_cache_fp8.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# FP8 simulated KV-cache quantization recipe for QARL.
+#
+# Use this same absolute path for policy.quant_cfg and
+# policy.generation.quant_cfg so Megatron and vLLM apply identical fake quant.
+metadata:
+  recipe_type: ptq
+  description: FP8 E4M3 simulated KV-cache quantization.
+quantize:
+  algorithm: max
```
Comment (sharonyu-115) [inline examples/modelopt/quant_configs/kv_cache_fp8.yaml:9]: @mxinO I wonder what is the amax calibration flow with modelopt. Is this done only once at step 0? Is the training data used for calibration?

Comment (mxinO) [reply examples/modelopt/quant_configs/kv_cache_fp8.yaml:9]: Currently, it only run ptq before training using specified calibration data, the training data won't be used, because it only has prompts.

File: docs/guides/quantization-aware-rl.md (hunk 1 above)
Comment (mxinO) [reply docs/guides/quantization-aware-rl.md:69]: Thanks for asking, I made a too broad description here, actually only one case is special currently, user can turn on kv cache quant only when use fp8 constant amax mode, when the quant scale will set to 1 automatically, and amax is not set. I will add constant amax for nvfp4 later if found helpful.

File: docs/assets/qarl_kv_cache_validation_accuracy.png
Comment (mxinO) [reply docs/assets/qarl_kv_cache_validation_accuracy.png:1]: yea, this is not a recipe for convergence now, the simulated kv cache is useful for 3 cases,
1. For accuracy recovery. After ptq, run an extra QARL stage to recover the accuracy. This is currently the main purpose. 
2. For large models (e.g. x hundreds billons parameters model), in which case the nvfp4 kv cache has less degradation. This will need further tests. 
3. Also enables recipe study, e.g. mixed quantization format, like only quantize some layers.

Comment (mxinO) [reply docs/assets/qarl_kv_cache_validation_accuracy.png:1]: Will add a comment in the doc.

File: examples/modelopt/quant_configs/kv_cache_fp8.yaml (hunk 3 above)
Comment (sharonyu-115) [reply examples/modelopt/quant_configs/kv_cache_fp8.yaml:9]: I see. That might be a reason why the training runs showed that constant amax works better. Thanks for explanation.

**Review by sharonyu-115 — APPROVED** [review]
